lib/passport-configurator.js

```diff
@@ -65,6 +67,31 @@ PassportConfigurator.prototype.setupModels = function(options) {
   if (!this.userCredentialModel.relations.user) {
     this.userCredentialModel.belongsTo(this.userModel, {as: 'user'});
   }
+  if (options.additionalUserModels) {
+    _.each(_.keys(options.additionalUserModels), (option) => {
+      let sOption = option.split('Model');
+      if (sOption.length === 2 && sOption[1] === '') {
+        let modelType = sOption[0].charAt(0).toUpperCase() +
+          sOption[0].substring(1).toLowerCase();
+        this[option] = options.additionalUserModels[option] ||
+          loopback.getModelByType(this.app.models[modelType]);
+        if (!this[option].relations.user) {
+          this[option].belongsTo(this.userModel, {as: 'user'});
+        }
+      }
+    });
+  }
```

Comment:
I'm confused here. The code seems to allow two cases:
```js
{
  myUserModel: MyUserModel,  // use `MyUserModel`
  yourUserModel: null (or false),  // use `Youruser` as the key to look up a model class
}
```
If my understanding is correct, please give an example in the jsdoc as it's not obvious to figure out.

Reply:
BTW, `yourUserModel` should map to `YourUser` instead of `Youruser`, right?

Reply:
I don't see these two properties are covered by any tests you added.

Reply:
@raymondfeng I utilized the same logic used for the userModel, userCredentialModel, and userIdentityModel code. And additional user models would map to their name in the config.
`{
  ...,
  additionalUserModels: {
    someUserRelatedModel: MyRelatedModel
  }
}`
This above example would map to this.someUserRelatedModel.

For the tests I am not that familiar with the mocha, so I looked at the current tests and did my best to replicate what was there. In the passport-configurator.test.js file I do see some options used for the setupModels call, but it only appears to be for ldap and not for any other strategies (facebook, oauth, saml, etc). Any additional models are optional but would require additional models, so I am not sure how to go about doing that with the current setup. Any recommendation would be appreciated.

Reply:
1. Can we extract the logic to derive the model name from a string into a utility function?
2. For the test, we can pass an `options` object with `additionalModels` and `additionalUserModels` and check if `this` contains such models.

Reply:
@raymondfeng 1. Where would you like to see that utility function? Should it be declared within the setupModels method, outside of any declared prototype methods, or somewhere else?
2. There currently does not exist any checks for userModel, userIdentityModel, or userCredentialModel that I can see. Should tests be added for them as well, and if so could you provide me with an example of how you would like those to be implemented so that I can base additional model tests off of them?
I am thinking something like this...
```
  it('has userModel, userCredentialModel, and userIdentityModel', function(done) {
    assert.equal(passportConfigurator.userModel, User, 'userModel should equal User');
    assert.equal(passportConfigurator.userCredentialModel, UserCredential,
      'userCredentialModel should equal UserCredential model');
    assert.equal(passportConfigurator.userIdentityModel, UserIdentity,
      'userIdentityModel should equal UserIdentity model');
    done();
  });
```
Currently these tests appear to work successfully. If you are satisfied with that, I will include them along with a similar test for additional models.

Reply:
@raymondfeng any input?